### ISSUE TITLE
cmd/govim: command-line mode to query "parent" cmd/govim instance

### ISSUE THREAD

**Issue by myitcv** [govim/govim#795]
A [recent exchange on Slack](https://gophers.slack.com/archives/CL8JDP1GD/p1581693138070200) discussed how to support the [LSP Symbol method](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#workspace_symbol), especially in combination with (fuzzy) finders like [`fzf`](https://github.com/junegunn/fzf). Using `fzf` as a popular example of a (fuzzy) finder, we explored what would be required to use `fzf` in conjunction with the Symbol method. 

The Symbol method takes a query parameter and returns the set of matching results. `fzf` supports using fuzzy finding based on some input, selecting the desired result from the list of results, then doing something with that selection, e.g. opening a file (in a (v)split)

The slightly wrinkle in this case is that we need to re-query the Symbol method as the user types in the fuzzy finder. We raised this in https://github.com/junegunn/fzf/issues/1883

As the response indicates, an approach similar to https://github.com/junegunn/fzf.vim/issues/948 is required here because `fzf` cannot call back into Vim in order to re-call the Symbol method. 

Hence this issue is a placeholder for making changes in `govim` to support a "child" `govim` instance querying a "parent" `govim` instance to, at least, call the Symbol method by some means. Gathering some of the notes from the Slack exchange:

* Offer a means by which `govim` can be run as a cmd to connect to an existing `govim` process and run certain queries/commands etc
* In `github.com/govim/govim/testdriver` we already provide a means by which we expose, via a command that connects to `govim`, a means to arbitrarily run Vim commands, evaluate expressions etc. This command is the special `vim` command that is used in the `testscript` tests
* A decent first step would be to refactor the `vim` command functionality in `testdriver` to be a part of `cmd/govim`, enabled by a special flag (e.g. `-parent SOCKET`). Hence a call to the Symbol method would look like:

```
govim -parent :4567 vim expr "GOVIMSymbol({\"query\": \"main\"})"
```

* We could then even bypass Vim and have `govim` directly relay certain calls to `gopls`, e.g.

```
govim -parent :4567 lsp Symbol "({\"query\": \"main\"})"
```

This would cut out the overhead in going `govim (command) -> govim -> Vim -> govim -> gopls` and back again, and instead make the sequence `govim (command) -> govim -> gopls` and back again; the hops via Vim are not required at all here

### Command API

`govim` would gain a new flag, `-parent`, which would take the Unix/TCP/IP socket to connect to. This flag would indicate the caller is the child, and expect commands 

At a bare minimum we need to expose the Symbol method, and it probably makes most sense to make that a direct query against `gopls`, not via Vim. 

Therefore it probably makes sense to create an `lsp` subcommand which takes a method name and a JSON encoded argument:

```
govim -parent :4567 lsp Symbol "({\"query\": \"main\"})"
```

We could also expose a `vim` subcommand in a similar way to the `testdriver` `vim` command used in `testscript` tests:

```
govim -parent :4567 vim expr getqflist()
```

### Outstanding questions

* Are there any potential security implications here, especially on shared machines? Because any process could conceivably connect to any `govim` instance in the scenario presented and query information from Vim, file contents etc, which might subvert file permissions. Does `gopls` not have the same problem?

### Related `gopls` issues

* [x/tools/gopls: support advanced query syntax for Symbol](https://github.com/golang/go/issues/37237)
* [x/tools/gopls: fully qualify symbol names for Symbol](https://github.com/golang/go/issues/38273)
* [x/tools/gopls: consider a -utf8 flag to talk UTF8 offsets](https://github.com/golang/go/issues/38274)

cc @findleyr @leitzler

**Comment by findleyr** [comment]
I've considered adding support for a `-session` flag to gopls commands, which would cause the command runner to connect to a specific existing gopls session. specifically so that editors can use command line invocation rather than jsonrpc2 when it is convenient. Would that help here, or would you rather run this through the govim command?

**Comment by myitcv** [comment]
@findleyr - to summarise our Slack exchange in response to this. I plan to expose these methods via `govim` because:

* this will allow arbitrary Vim functions (whether `govim` related or not) to be used as the reload command (see https://github.com/junegunn/fzf.vim/issues/948)
* this will also allow custom formatting of results. For example, the input to `fzf` needs to be `grep`-like line-wise: the raw output from `gopls` would be encoded JSON
* this will allow column offsets to be corrected to UTF8. All outputs from `gopls` are in terms of UTF16 code points (hence https://github.com/golang/go/issues/38274)

**Comment by myitcv** [comment]
I've just pushed a first cut of support for this in #848 (and will continue to force-push fixes etc). It's only tested with `fzf` for now and requires config like the following in your `.vimrc`. Please give it a try and provide feedback in this issue!

[![asciicast](https://asciinema.org/a/zPyOkQu8p2XxXKLkrOlLc7I8b.svg)](https://asciinema.org/a/zPyOkQu8p2XxXKLkrOlLc7I8b)

```vim
" With thanks and reference to github.com/junegunn/fzf.vim/issues/948 which
" inspired the following
function! s:handleSymbol(sym) abort
  " a:sym is a [2]string array where the first element is the
  " key pressed (or empty if simply Enter), and the second element
  " is the entry selected in fzf, i.e. the match.
  "
  " The match will be of the form:
  "
  "   $filename:$line:$col: $match
  "
  if len(a:sym) == 0
    return
  endif
  let l:cmd = get(s:symbolActions, a:sym[0], "")
  let l:match = a:sym[1]
  let l:parts = split(l:match, ":")
  execute 'silent' l:cmd
  execute 'buffer' bufnr(l:parts[0], 1)
  call cursor(l:parts[1], l:parts[2])
  normal! zz
endfunction

" s:symbolActions are the actions that, in addition to plain <Enter>,
" we want to be able to fire from fzf. Here we map them to the corresponding
" command in VimScript.
let s:symbolActions = {
  \ 'ctrl-t': 'tab split',
  \ 'ctrl-x': 'split',
  \ 'ctrl-v': 'vsplit',
  \ }

" GovimFZFSymbol is a user-defined function that can be called to start fzf in
" a mode whereby it uses govim's new child-parent capabilities to query the
" parent govim instance for gopls Symbol method results that then are used to
" drive fzf.
function GovimFZFSymbol()
  let l:expect_keys = join(keys(s:symbolActions), ',')
  let l:source = GOVIMParentCommand().' gopls Symbol -quickfix -rel'
  let l:reload = l:source." {q}"
  call fzf#run(fzf#wrap({
        \ 'source': source,
        \ 'sink*': function('s:handleSymbol'),
        \ 'options': [
        \       '-d', ':',
        \       '-n', '4',
        \       '--expect='.l:expect_keys,
        \       '--bind',
        \       'change:reload:'.l:reload
        \ ]}))
endfunction

" Bind GovimFZFSymbol to <Leader>s, where <Leader> is \ for me
nmap <Leader>s :call GovimFZFSymbol()<CR>
```

Some notes from this experiment:

* Displaying the results in `fzf` using the format `filename:line:col: ....` is ugly and hard to parse. Better still would be to display the fully qualified symbol. See next point...
* `gopls` only performs case-insensitive matches against the symbol name (e.g. `main`) rather than fuzzy search against the fully qualified symbol (e.g. `example.com/blah.main`). Furthermore only the symbol name is returned in the `[]SymbolInformation` results. There is a further question of whether the kind and type of the symbol should be considered in the fuzzy matcher in the `Symbol` method; the type should almost certainly be available to the client for display purposes. Raised https://github.com/golang/go/issues/38273 for discussion
* `gopls` searching the main module and its transitive dependencies is not a sane default. Most of the time I want to be finding symbols in my code (the main module). On the rare occasion that I want to be searching the main module's transitive dependencies I will perform such a specific search. Raised https://github.com/golang/go/issues/37237 for discussion
* Because `gopls` speaks the LSP protocol, matches returned by the `Symbol` method are in terms of UTF16 code point offsets. This means that the parent `govim` instance has to translate all offsets in returned results before returning the results to the child. Whilst this is possible, it's less than desirable. Raised https://github.com/golang/go/issues/38274 for discussion (albeit a lower priority)

**Comment by myitcv** [comment]
I've just force-pushed a further change to #848. This includes:

* requiring `x/tools` and `x/tools/gopls` based on https://go-review.googlesource.com/c/tools/+/227677, which starts to flesh out a more advanced query syntax:
  * the fully qualified symbol is now searched against (https://github.com/golang/go/issues/38273)
  * uses server-side fuzzy searching 
  * defines and implements boolean search query options `workspaceOnly` and `nonWorkspaceExportedOnly`. The former constrains the search to packages in the workspace (main module); the latter constrains the search to exported identifiers in packages outside of the workspace (main module). To specify such an option include in the query as follows: `workspaceOnly::true main`
* improved UI/UX through use of the fully qualified symbols
* slight tweak to the suggested `.vimrc` config:

```vim
" With thanks and reference to github.com/junegunn/fzf.vim/issues/948 which
" inspired the following
function! s:handleSymbol(sym) abort
  " a:sym is a [2]string array where the first element is the
  " key pressed (or empty if simply Enter), and the second element
  " is the entry selected in fzf, i.e. the match.
  "
  " The match will be of the form:
  "
  "   $filename:$line:$col: $match
  "
  if len(a:sym) == 0
    return
  endif
  let l:cmd = get(s:symbolActions, a:sym[0], "")
  let l:match = a:sym[1]
  let l:parts = split(l:match, ":")
  execute 'silent' l:cmd
  execute 'buffer' bufnr(l:parts[0], 1)
  call cursor(l:parts[1], l:parts[2])
  normal! zz
endfunction

" s:symbolActions are the actions that, in addition to plain <Enter>,
" we want to be able to fire from fzf. Here we map them to the corresponding
" command in VimScript.
let s:symbolActions = {
  \ 'ctrl-t': 'tab split',
  \ 'ctrl-x': 'split',
  \ 'ctrl-v': 'vsplit',
  \ }

" GovimFZFSymbol is a user-defined function that can be called to start fzf in
" a mode whereby it uses govim's new child-parent capabilities to query the
" parent govim instance for gopls Symbol method results that then are used to
" drive fzf.
function GovimFZFSymbol(queryAddition)
  let l:expect_keys = join(keys(s:symbolActions), ',')
  let l:source = GOVIMParentCommand().' gopls Symbol -quickfix '.a:queryAddition
  let l:reload = l:source." {q}"
  call fzf#run(fzf#wrap({
        \ 'source': source,
        \ 'sink*': function('s:handleSymbol'),
        \ 'options': [
        \       '--with-nth', '2..',
        \       '--expect='.l:expect_keys,
        \       '--phony',
        \       '--bind', 'change:reload:'.l:reload
        \ ]}))
endfunction

nmap <Leader>s :call GovimFZFSymbol('nonWorkspaceExportedOnly::true workspaceOnly::true')<CR>
nmap <Leader>S :call GovimFZFSymbol('nonWorkspaceExportedOnly::true workspaceOnly::false')<CR>
```

Note the two bindings at the end demonstrate how to auto-apply search query options: the former (`\s`) constrains search to the workspace (main module). The latter (`\S`) does not.

**Comment by myitcv** [comment]
I've just force-pushed a further change to #848. This includes:

* requiring `x/tools` and `x/tools/gopls` based on the latest patchset to https://go-review.googlesource.com/c/tools/+/227677/24, which starts to flesh out a more advanced query syntax:
  * the fully qualified symbol is now searched against (https://github.com/golang/go/issues/38273)
  * uses server-side fuzzy searching 
  * defines and implements search query options:
    * `workspaceOnly::(true|false)` - constrain the search to workspace (main module) packages 
 - default `false`
    * `nonWorkspaceExportedOnly::(true|false)` - constrain the search to exported identifiers in packages outside of the workspace (main module) - default `false`
    * `structFields::(true|false)` - whether to include `struct` type fields in the search results 
 - default `true`
    * `interfaceMethods::(true|false)` - whether to include `interface` type methods in the search results - default `true`
    * `kind::X,Y,Z,...` - the symbol kinds to consider when search. This is a comma-separate list of integer values corresponding to `protocol.SymbolKind` - default is all symbol kinds
    * `packageOnly::file:///path/to/file.go` - constrain the search to the package that contains `/path/to/file.go` - default is to not constrain by package
* improved UI/UX through use of the fully qualified symbols
* requirements still the same as before:
  * https://github.com/junegunn/fzf
  * https://github.com/junegunn/fzf.vim
* slight tweak to the suggested `.vimrc` config:

```vim
" With thanks and reference to github.com/junegunn/fzf.vim/issues/948 which
" inspired the following
function! s:handleSymbol(sym) abort
  " a:sym is a [2]string array where the first element is the
  " key pressed (or empty if simply Enter), and the second element
  " is the entry selected in fzf, i.e. the match.
  "
  " The match will be of the form:
  "
  "   $filename:$line:$col: $match
  "
  if len(a:sym) == 0
    return
  endif
  let l:cmd = get(s:symbolActions, a:sym[0], "")
  let l:match = a:sym[1]
  let l:parts = split(l:match, ":")
  execute 'silent' l:cmd
  execute 'buffer' bufnr(l:parts[0], 1)
  call cursor(l:parts[1], l:parts[2])
  normal! zz
endfunction

" s:symbolActions are the actions that, in addition to plain <Enter>,
" we want to be able to fire from fzf. Here we map them to the corresponding
" command in VimScript.
let s:symbolActions = {
  \ 'ctrl-t': 'tab split',
  \ 'ctrl-x': 'split',
  \ 'ctrl-v': 'vsplit',
  \ }

" GovimFZFSymbol is a user-defined function that can be called to start fzf in
" a mode whereby it uses govim's new child-parent capabilities to query the
" parent govim instance for gopls Symbol method results that then are used to
" drive fzf.
function GovimFZFSymbol(queryAddition)
  let l:commonOptions = 'structFields::false interfaceMethods::false'
  let l:expect_keys = join(keys(s:symbolActions), ',')
  let l:source = join(GOVIMParentCommand(), " ").' gopls Symbol -quickfix '.l:commonOptions.' '.a:queryAddition
  let l:reload = l:source." {q}"
  call fzf#run(fzf#wrap({
        \ 'source': l:source,
        \ 'sink*': function('s:handleSymbol'),
        \ 'options': [
        \       '--with-nth', '2..',
        \       '--expect='.l:expect_keys,
        \       '--phony',
        \       '--tac',
        \       '--bind', 'change:reload:'.l:reload
        \ ]}))
endfunction

nmap <Leader>s :call GovimFZFSymbol('nonWorkspaceExportedOnly::true packageOnly::file://'.expand('%:p'))<CR>
nmap <Leader>S :call GovimFZFSymbol('nonWorkspaceExportedOnly::true workspaceOnly::true')<CR>
nmap <Leader>g :call GovimFZFSymbol('nonWorkspaceExportedOnly::true workspaceOnly::false')<CR>
nmap <Leader>G :call GovimFZFSymbol('nonWorkspaceExportedOnly::false workspaceOnly::false')<CR>
```

Note the bindings at the end demonstrate how to auto-apply search query options: 

* `\s` is a `packageOnly`
* `\S` is a `workspaceOnly` search
* `\g` searches symbols in all packages transitively reachable from workspace packages, but limits the results to exported-only values in non-workspace packages
* `\G`searches symbols in all packages transitively reachable from workspace packages

**Comment by myitcv** [comment]
Updated screencast: https://asciinema.org/a/Vqr3EgsRlMq4y8MNJZctZgIE4

**Comment by myitcv** [comment]
I've just force-pushed a further change to #848, mainly to rebase the branch against `master` which has new changes merged in, including an update to the latest `master` of `x/tools` and `x/tools/gopls`. There is no change in the config required or functionality described in https://github.com/govim/govim/issues/795#issuecomment-616588901.

**Comment by myitcv** [comment]
Thanks to some sterling work by @findleyr, we now have a revised version of this pushed to https://github.com/govim/govim/pull/848 that is based on `gopls` tip.

For details on how to use this with `fzf` please see https://github.com/govim/govim/wiki/vimrc-tips#-tip-use-fzf-with-workspace-symbol-search

Note that there are now two configuration options that drive symbol search:

* `SymbolMatcher` - whether to use `fuzzy`, `casesensitive` or `caseinsensitive` matching (fuzzy is the default)
* `SymbolStyle` - how to qualify the search results. `full` fully qualifies each symbol with the package import path, `dynamic` minimally qualifies each candidate so as to give the best score, `package` only qualifies a candidate with the package name

Latest screencast: https://asciinema.org/a/yEXLJwIsHLTLD0Fy9k5jHyVbE
